---
title: "cfDNA Analysis - An Examination of the DELFI Approach"
author: "Francois Collin"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  BiocStyle::html_document:
    code_folding: hide
    toc: true
    # does this have an effect
    fig_caption: yes    
    # this has no effect
    number_sections: yes
    css: /Users/fcollin/Documents/Projects/_pandocFiles/pandoc3.css
bibliography: [../bib/cfDNA.bib, ../bib/bibFile.bib]
#csl: ../../bibFiles/acm-sig-proceedings.csl  - doest work with pandoc-siteproc
csl: ../csl/cell-numeric.csl
#biblio-style: acm
link-citations: true
vignette: >
 %\VignetteEncoding{UTF-8}
---

<!--
<style type="text/css">
Proteomics Data Analysis - A Case Study with a NSCLC Dataset"
body{ /* Normal  */ 
 font-size: 16px; 
}
td {  /* Table  */ 
 font-size: 12; 
}
h1.title { 
 font-size: 28px; color: DarkGreen; 
}
h1 { /* Header 1 */ 
 font-size: 24px; color: DarkBlue; 
}
h2 { /* Header 2 */ 
 font-size: 18px; color: DarkBlue; 
}
h3 { /* Header 3 */ 
 font-size: 18px;
 font-family: "Times New Roman", Times, serif;
 color: DarkBlue;
}
code.r{ /* Code block */
    font-size: 12px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>
-->


```{r m1a-GlobalOptions, results="hide", include=FALSE, cache=FALSE}
knitr::opts_knit$set(stop_on_error = 2L) #really make it stop
options(knitr.table.format = 'html')

#knitr::dep_auto()
```
<!-- ######################################################################## -->


```{r m1a-Prelims, include=FALSE, echo=FALSE, results='hide', message=FALSE} 

 FN <- "_M1A-exploreDELFI"
if(sum(grepl(FN, list.files()))==0) stop("Check FN")

 suppressMessages(require(rmarkdown))
 suppressMessages(require(knitr))
 options(stringsAsFactors=F) 


 suppressPackageStartupMessages(require(edgeR))

 suppressPackageStartupMessages(require(methods))
 suppressPackageStartupMessages(require(bookdown))

 suppressPackageStartupMessages(require(data.table))
 options(datatable.fread.datatable=F)

 suppressPackageStartupMessages(require(plyr))
 suppressPackageStartupMessages(require(dplyr))
 suppressPackageStartupMessages(require(magrittr))

 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))


 WRKDIR <- '..'
 if(!file.exists(WRKDIR)) stop("WRKDIR ERROR", WRKDIR)

 # do once

 # Shotcuts for knitting and redering while in R session
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste('',FN,".html", sep=''))

 rr <- function(n='') render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep=''), output_dir='Scripts')

 bb <- function(n='') browseURL(paste('',FN,".html", sep=''))

 # The usual shorcuts
 zz <- function(n='') source(paste('', "t", n, sep=''))

 # file rmarkdown file management options: cache, figures
 cache_DIR <- file.path(WRKDIR, 'Scripts', 'cache/M1A/')
 suppressMessages(dir.create(cache_DIR, recursive=T))
 opts_chunk$set(cache.path=cache_DIR)

 figures_DIR <- file.path(WRKDIR, 'Scripts', 'figures/M1A/')
 suppressMessages(dir.create(figures_DIR, recursive=T))
 opts_chunk$set(fig.path=figures_DIR)

 #tables_DIR <- file.path(WRKDIR, 'Scripts', 'tables/M1A/')
 #suppressMessages(dir.create(table_DIR, recursive=T))
 #opts_chunk$set(fig.path=table_DIR)
 
 # need a local copy of help_DIR
 #help_DIR <- file.path(WRKDIR, 'help_files')
 help_DIR <- file.path('.', 'help_files')
 suppressMessages(dir.create(help_DIR, recursive=T))
 
 temp_DIR <- file.path(WRKDIR, 'temp_files')
 suppressMessages(dir.create(temp_DIR, recursive=T))

```
<!-- ######################################################################## -->


*** 
```{r m1a-utilityFns, echo=FALSE}
 # Here we define some utility functions
source('utilityFns.r')

```
<!-- ######################################################################## -->


***

# Synopsis 

This script explores the data reported on in
Cristiano et al. (2020) [@Cristiano:2019aa]
These data are part of the supporting information
available through the
[online version of the paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6774252/).

<!--
Questions of interest:

* What are the primary sources of variability?

* What is the relative level of biological versus technical variability
in proteomic data?

* In this particular dataset, how do various classification models
perform in terms of cross-validated and test set performance?
-->

Some findings that stand out:

TBD


***


# Read in data

```{r m1a-read-table-decr, include=F}
DD <- 'Cristiano'

# Sample meta data
DD_META_FILE <- '../Refs/Cristiano_supp/NIHMS1529141-supplement-2.xlsx'
if(!file.exists(DD_META_FILE))
stop("Specify DD_META_FILE")

DD.metaSheets.vec <- readxl::excel_sheets(DD_META_FILE)

DD.metaSheetsHdr.vec <- do.call('c', lapply(DD.metaSheets.vec,
function(SHEET) readxl::read_xlsx(DD_META_FILE, sheet=SHEET, n_max=1)[1,1]))
```

```{r m1a-read-table-decr-2}

cat("Sheets found in", DD_META_FILE, '\n')
names(DD.metaSheetsHdr.vec)

```

<br/>

## read `r names(DD.metaSheetsHdr.vec)[1]`


```{r m1a-readSampDesc, eval=T, cache=T, cache.vars=c('DD', 'DD.sampDesc.frm'),warning=F}
DD.sampDesc.frm <- as.data.frame(
 readxl::read_xlsx(DD_META_FILE, sheet=DD.metaSheets.vec[1], skip=1))

names(DD.sampDesc.frm) <- 
sub("Patient Type", "patientType",
sub("Sample Type", "sampleType",
sub("Age at Diagnosis", "Age",
sub("TNM Staging", "TMM_Staging",                            
sub("Site of Primary Tumor", "primarySite",                  
sub("Histopathological Diagnosis", "histoDx",            
sub("Degree of Differentiation", "degDiff",              
sub("Location of Metastases at Diagnosis", "metLoc",
sub("Volume of Plasma \\(ml\\)", "plasmaVol",                  
sub("cfDNA Extracted \\(ng/ml\\)", "cfDNA_Extracted",                
sub("cfDNA Input \\(ng/ml\\)", "cfDNA_Input",                    
sub("Whole Genome Fragment Profile  Analysis", "WGFP",
sub("Targeted Fragment Profile Analysis", "TGFP",     
sub("Targeted Mutation Analysis", "TGMT",
names(DD.sampDesc.frm)))))))))))))))

DD.sampDesc.frm$Age <- as.numeric(DD.sampDesc.frm$Age)
DD.sampDesc.frm$plasmaVol <- as.numeric(DD.sampDesc.frm$plasmaVol)
DD.sampDesc.frm$cfDNA_Extracted <- as.numeric(DD.sampDesc.frm$cfDNA_Extracted)
DD.sampDesc.frm$cfDNA_Input <- as.numeric(DD.sampDesc.frm$cfDNA_Input)

DD.sampDesc.frm$Group <- sapply(strsplit(DD.sampDesc.frm$patientType, 
   split=' '), '[', 1)

DD.sampDesc.frm$GroupF <- factor(DD.sampDesc.frm$Group,
  levels=c("Healthy", setdiff(sort(unique(DD.sampDesc.frm$Group)), "Healthy")))

# Order by Group, gender, age
o.v <- with(DD.sampDesc.frm, order(GroupF, Gender, Age))
DD.sampDesc.frm <- DD.sampDesc.frm[o.v,]

saveObj(paste0(DD, '.sampDesc.frm'), 'DD.sampDesc.frm')

DT::datatable(DD.sampDesc.frm)

```


Silently get Annotation...

<br/>

```{r m1a-annotation, cache=F}
# Annotation
#################
# kelly's colors - https://i.kinja-img.com/gawker-media/image/upload/1015680494325093012.JPG
# https://gist.github.com/ollieglass/f6ddd781eeae1d24e391265432297538
# KellyColors.vec <-  see web site
# REMOVED '#F2F3F4' in first entry
KellyColors.vec <- c(
  "#222222", "#F3C300", "#875692", "#F38400", "#A1CAF1",
  "#BE0032", "#C2B280", "#848482", "#008856", "#E68FAC", "#0067A5",
  "#F99379", "#604E97", "#F6A600", "#B3446C", "#DCD300", "#882D17",
  "#8DB600", "#654522", "#E25822", "#2B3D26"
)
col_vector <- KellyColors.vec


# Group Annotation
GroupLegend.vec <- sort(unique(DD.sampDesc.frm$Group))
GroupCol.vec <- col_vector[1:length(GroupLegend.vec)]
names(GroupCol.vec) <- GroupLegend.vec

GroupPch.vec <- 1:length(GroupLegend.vec)
names(GroupPch.vec) <- GroupLegend.vec

# Gender Annotation
GenderLegend.vec <- sort(unique(DD.sampDesc.frm$Gender))
GenderCol.vec <- col_vector[1:length(GenderLegend.vec)]
names(GenderCol.vec) <- GenderLegend.vec

GenderPch.vec <- 1:length(GenderLegend.vec)
names(GenderPch.vec) <- GenderLegend.vec

# Site Annotation - DNA


```

<br/>

Look at plasma attribute.


```{r m1a-plasma-summaries, fig.cap="Plasma Summaries"}

tmp <- data.frame(DD.sampDesc.frm %>% dplyr::select(plasmaVol:cfDNA_Input))

knitr::kable(t(apply(tmp, 2, summary)),
  digits=2, caption="Plasma Summaries") %>%
 kableExtra::kable_styling(full_width = F)

```

<br/>


```{r m1a-sample-counts, fig.cap="Sample Counts by Type"}

knitr::kable(
 with(DD.sampDesc.frm, table(patientType, sampleType)),
 caption="Sample Counts") %>%
 kableExtra::kable_styling(full_width = F)

```


```{r m1a-scatter-plasma-by-group, fig.height=6, fig.width=11, fig.cap="Plasma Characteristics", include=F, cache=T, eval=F, echo=F}

 with(DD.sampDesc.frm, 
  plot(cfDNA_Extracted, cfDNA_Input, log='xy',
       col=GroupCol.vec[DD.sampDesc.frm$Group],
       pch=GroupPch.vec[DD.sampDesc.frm$Group]))

```

<br/>

```{r m1a-boxplot-plasma-by-group, fig.height=6, fig.width=11, fig.cap="Plasma Characteristics", cahce=T, cache.vars=''}


 cfDNA_Input.lst <- with(DD.sampDesc.frm, split(cfDNA_Input, GroupF))

 boxplot(cfDNA_Input.lst, log='y',
  add = F,
  #ylim = c(2, 9), 
  staplewex = 0, # remove horizontal whisker lines
  staplecol = "white", # just to be totally sure
  outline = T, # DO NOT remove outlying points
  whisklty = 0, # remove vertical whisker lines
  las = 2, horizontal = F, xaxt = "n",
  border = GroupCol.vec[names(cfDNA_Input.lst)])

 abline(h=median(cfDNA_Input.lst[["Healthy"]], na.rm=T), col=GroupCol.vec["Healthy"])

 title("cfDNA_Input by Group")

axis(side=1, at=1:length(cfDNA_Input.lst), 
labels=names(cfDNA_Input.lst), las=2)

text(x=1:length(cfDNA_Input.lst), y=rep(0.5, length(cfDNA_Input.lst)),
  labels=sapply(cfDNA_Input.lst, length))


```

<br/>



## read `r names(DD.metaSheetsHdr.vec)[4]`

Note: We had to separate 
`r names(DD.metaSheetsHdr.vec)[3]` into
`r names(DD.metaSheetsHdr.vec)[4]` and
`r names(DD.metaSheetsHdr.vec)[5]` to simplify reading in
the data.  Here we read the two separately and then 
recombine them.

```{r m1a-readTargFragWt, eval=T, cache=T, cache.vars=c('DD.targFragWt.frm'),warning=F} 

DD.targFragWt.frm <- as.data.frame(
 readxl::read_xlsx(DD_META_FILE, sheet=DD.metaSheets.vec[4], skip=1))

names(DD.targFragWt.frm) <- 
sub( "Patient Type", "patientType",
sub( "Stage at Diagnosis", "stageAtDx",
sub( "Alteration Type", "altType",
sub( "Amino Acid \\(Protein\\)", "aminoAcid",
sub( "Mutation Type", "mtType",
sub( "Hotspot Alteration", "hotspotAlt", 
sub( "Alteration Detected in Tissue", "altDetectTissue",
sub( "Mutant Allele Fraction", "mtAllFrac",
sub( "Distinct Coverage", "WtDistinctCov",
sub( "Minimum cfDNA Fragment Size \\(bp\\)", "WtFragSizeMin",
sub( "25th Percentile cfDNA Fragment Size \\(bp\\)", "WtFragSizeQ1",
sub( "Mode cfDNA Fragment Size \\(bp\\)", "WtFragSizeMode",
sub( "Median cfDNA Fragment Size \\(bp\\)", "WtFragSizeMedian",
sub( "Mean cfDNA Fragment Size \\(bp\\)", "WtFragSizeMean",
sub( "75th Percentile cfDNA Fragment Size \\(bp\\)", "WtFragSizeQ3",
sub( "Maximum cfDNA Fragment Size \\(bp\\)", "WtFragSizeMax",
sub( "Difference between Median Mutant and Wild-type cfDNA Fragment Sizes \\(bp\\)", "diffMedianSizeMtWt",
sub( "Difference between Mean Mutant and Wild-type cfDNA Fragment Sizes \\(bp\\)", "diffMeanSizeMtWt",
sub( "Adjusted P Value of Difference between Mutant and Wild-type cfDNA Fragment Sizes", "adjPvalue",
names(DD.targFragWt.frm)))))))))))))))))))) 



DD.targFragWt.frm$Group <- sapply(strsplit(DD.targFragWt.frm$patientType, 
   split=' '), '[', 1)

DD.targFragWt.frm$GroupF <- factor(DD.targFragWt.frm$Group)
## No Healthy here
  #levels=c("Healthy", setdiff(sort(unique(DD.targFragWt.frm$Group)), "Healthy")))


```

<br/>

## read `r names(DD.metaSheetsHdr.vec)[5]`



```{r m1a-readTargFragMt, eval=T, cache=T, cache.vars=c('DD.targFragMt.frm'),warning=F,include=F} 

DD.targFragMt.frm <- as.data.frame(
 readxl::read_xlsx(DD_META_FILE, sheet=DD.metaSheets.vec[5], skip=1))

names(DD.targFragMt.frm) <-
sub( "Patient Type", "patientType",
sub( "Stage at Diagnosis", "stageAtDx",
sub( "Alteration Type", "altType",
sub( "Amino Acid \\(Protein\\)", "aminoAcid",
sub( "Mutation Type", "mtType",
sub( "Hotspot Alteration", "hotspotAlt",
sub( "Alteration Detected in Tissue", "altDetectTissue",
sub( "Mutant Allele Fraction", "mtAllFrac",
sub( "Distinct Coverage", "MtDistinctCov",
sub( "Minimum cfDNA Fragment Size \\(bp\\)", "MtFragSizeMin",
sub( "25th Percentile cfDNA Fragment Size \\(bp\\)", "MtFragSizeQ1",
sub( "Mode cfDNA Fragment Size \\(bp\\)", "MtFragSizeMode",
sub( "Median cfDNA Fragment Size \\(bp\\)", "MtFragSizeMedian",
sub( "Mean cfDNA Fragment Size \\(bp\\)", "MtFragSizeMean",
sub( "75th Percentile cfDNA Fragment Size \\(bp\\)", "MtFragSizeQ3",
sub( "Maximum cfDNA Fragment Size \\(bp\\)", "MtFragSizeMax",
sub( "Difference between Median Mutant and Wild-type cfDNA Fragment Sizes \\(bp\\)", "diffMedianSizeMtWt",
sub( "Difference between Mean Mutant and Wild-type cfDNA Fragment Sizes \\(bp\\)", "diffMeanSizeMtWt",
sub( "Adjusted P Value of Difference between Mutant and Wild-type cfDNA Fragment Sizes", "adjPvalue",
names(DD.targFragMt.frm))))))))))))))))))))


DD.targFragMt.frm$Group <- sapply(strsplit(DD.targFragWt.frm$patientType, 
   split=' '), '[', 1)
DD.targFragMt.frm$GroupF <- factor(DD.targFragMt.frm$Group)


```

<br/>

## Merge Tables 3A and 3B


```{r m1a-mergeTargFragWtMt, eval=T, cache=T, cache.vars=c('DD.targFragWtMt.frm'),warning=F,fig.cap="Wild-type and Mutant Fragments Data",include=T} 

DD.targFragWtMt.frm <- merge(DD.targFragWt.frm,DD.targFragMt.frm)

with(DD.targFragWtMt.frm, table(GroupF))

# samples per Patient
tmp.tbl <- with(DD.targFragWtMt.frm, table(table(Patient)))
tmp.frm <- data.frame(N_Patient=as.vector(tmp.tbl), Count=names(tmp.tbl))
knitr::kable(tmp.frm,
 caption="Number of samples per Patient") %>%
 kableExtra::kable_styling(full_width = F)


cat("samples per Patient by Group\n")
lapply(split(DD.targFragWtMt.frm, DD.targFragWtMt.frm$Group), 
function(GRP.targFragWtMt.frm) {
tmp.tbl <- with(GRP.targFragWtMt.frm, table(table(Patient)))
tmp.frm <- data.frame(N_Patient=as.vector(tmp.tbl), Count=names(tmp.tbl))
tmp.frm})


# Order by Group, gender, age
o.v <- with(DD.targFragWtMt.frm, order(GroupF, stageAtDx, Patient))
DD.targFragWtMt.frm <- DD.targFragWtMt.frm[o.v,]

saveObj(paste0(DD, '.targFragWtMt.frm'), 'DD.targFragWtMt.frm')

DT::datatable(DD.targFragWtMt.frm)

```

<br/>

# Examine the Fragment Size Data

## Look at fragSizeMedian, by Group

```{r m1a-fragSizeMedian-by-Group, cache=T, cache.vars='',fig.height=5, fig.width=11,fig.cap="fragSizeMedian by Group"}

WtFragSizeMedian.lst <- with(DD.targFragWtMt.frm, split(WtFragSizeMedian, Group))
names(WtFragSizeMedian.lst) <- paste0(names(WtFragSizeMedian.lst), '-Wt')

MtFragSizeMedian.lst <- with(DD.targFragWtMt.frm, split(MtFragSizeMedian, Group))
names(MtFragSizeMedian.lst) <- paste0(names(MtFragSizeMedian.lst), '-Mt')

WtMtFragSizeMedian.lst <- c(WtFragSizeMedian.lst, MtFragSizeMedian.lst)
WtMtFragSizeMedian.lst <- WtMtFragSizeMedian.lst[order(names(WtMtFragSizeMedian.lst))]

par(mar=par("mar")+c(2,0,0,0))
boxplot(WtMtFragSizeMedian.lst, log='y', las=2,
 border=ifelse(grepl('Wt', names(WtMtFragSizeMedian.lst)), 'blue', 'red'))

title("fragSizeMedian by Group")

```

<br/>

## Look at within Vs between patient variability - fragSizeMedian

In the following figure, the different data points within
each patient correspond to different loci or gene mutations.  See
Figure \@ref(fig:m1a-mergeTargFragWtMt) above.

```{r m1a-fragSizeMedian-by-Group-2, cache=T, cache.vars='',fig.height=5, fig.width=11,fig.cap="fragSizeMedian by Group"}

WtFragSizeMedian.lst <- with(DD.targFragWtMt.frm, split(WtFragSizeMedian, Group))
MtFragSizeMedian.lst <- with(DD.targFragWtMt.frm, split(MtFragSizeMedian, Group))
Patient.lst <- with(DD.targFragWtMt.frm, split(Patient, Group))

par(mar=par("mar")+c(2,0,0,0))

for(GRP in names(WtFragSizeMedian.lst)){
 Grp.WtFragSizeMedian.vec <- WtFragSizeMedian.lst[[GRP]]
 Grp.MtFragSizeMedian.vec <- MtFragSizeMedian.lst[[GRP]]
 Grp.Patient.vec <- Patient.lst[[GRP]]

 Grp.WtFragSizeMedian.lst <- split(Grp.WtFragSizeMedian.vec, Grp.Patient.vec)
 Grp.MtFragSizeMedian.lst <- split(Grp.MtFragSizeMedian.vec, Grp.Patient.vec)

 boxplot(Grp.WtFragSizeMedian.lst, las=2, col=0, border=0, outline=F, ylim=c(145,195))
 abline(v=1:length(Grp.WtFragSizeMedian.lst)-0.5, col='grey')
 # Wt
 for(JJ in 1:length(Grp.WtFragSizeMedian.lst))
 points(jitter(rep(JJ, length(Grp.WtFragSizeMedian.lst[[JJ]])), amount=.25),
        Grp.WtFragSizeMedian.lst[[JJ]], 
        col='blue', pch=1:length(Grp.WtFragSizeMedian.lst[[JJ]]))
 # Mt 
 for(JJ in 1:length(Grp.MtFragSizeMedian.lst))
 points(jitter(rep(JJ, length(Grp.MtFragSizeMedian.lst[[JJ]])), amount=.25),
        Grp.MtFragSizeMedian.lst[[JJ]], 
        col='red', pch=1:length(Grp.MtFragSizeMedian.lst[[JJ]]))

 title(paste("Median Fragment Size -", GRP, "\nBlue=Wt,  Red=Mt,  Pch=Locus (within Patient)"))
}
 

```


<br/>

## Look at fragSizeMean, by Group

```{r m1a-fragSizeMean-by-Group, cache=T, cache.vars='',fig.height=5, fig.width=11,fig.cap="fragSizeMean by Group"}

WtFragSizeMean.lst <- with(DD.targFragWtMt.frm, split(WtFragSizeMean, Group))
names(WtFragSizeMean.lst) <- paste0(names(WtFragSizeMean.lst), '-Wt')

MtFragSizeMean.lst <- with(DD.targFragWtMt.frm, split(MtFragSizeMean, Group))
names(MtFragSizeMean.lst) <- paste0(names(MtFragSizeMean.lst), '-Mt')

WtMtFragSizeMean.lst <- c(WtFragSizeMean.lst, MtFragSizeMean.lst)
WtMtFragSizeMean.lst <- WtMtFragSizeMean.lst[order(names(WtMtFragSizeMean.lst))]

par(mar=par("mar")+c(2,0,0,0))
boxplot(WtMtFragSizeMean.lst, log='y', las=2,
 border=ifelse(grepl('Wt', names(WtMtFragSizeMean.lst)), 'blue', 'red'))

title("fragSizeMean by Group")

```

<br/>

## Look at within Vs between patient variability - fragSizeMean


```{r m1a-fragSizeMean-by-Group-2, cache=T, cache.vars='',fig.height=5, fig.width=11,fig.cap="fragSizeMean by Group"}

WtFragSizeMean.lst <- with(DD.targFragWtMt.frm, split(WtFragSizeMean, Group))
MtFragSizeMean.lst <- with(DD.targFragWtMt.frm, split(MtFragSizeMean, Group))
Patient.lst <- with(DD.targFragWtMt.frm, split(Patient, Group))

par(mar=par("mar")+c(2,0,0,0))

for(GRP in names(WtFragSizeMean.lst)){
 Grp.WtFragSizeMean.vec <- WtFragSizeMean.lst[[GRP]]
 Grp.MtFragSizeMean.vec <- MtFragSizeMean.lst[[GRP]]
 Grp.Patient.vec <- Patient.lst[[GRP]]

 Grp.WtFragSizeMean.lst <- split(Grp.WtFragSizeMean.vec, Grp.Patient.vec)
 Grp.MtFragSizeMean.lst <- split(Grp.MtFragSizeMean.vec, Grp.Patient.vec)

 boxplot(Grp.WtFragSizeMean.lst, las=2, col=0, border=0, outline=F, ylim=c(145,220))
 abline(v=1:length(Grp.WtFragSizeMean.lst)-0.5, col='grey')
 # Wt
 for(JJ in 1:length(Grp.WtFragSizeMean.lst))
 points(jitter(rep(JJ, length(Grp.WtFragSizeMean.lst[[JJ]])), amount=.25),
        Grp.WtFragSizeMean.lst[[JJ]], 
        col='blue', pch=1:length(Grp.WtFragSizeMean.lst[[JJ]]))
 # Mt 
 for(JJ in 1:length(Grp.MtFragSizeMean.lst))
 points(jitter(rep(JJ, length(Grp.MtFragSizeMean.lst[[JJ]])), amount=.25),
        Grp.MtFragSizeMean.lst[[JJ]], 
        col='red', pch=1:length(Grp.MtFragSizeMean.lst[[JJ]]))

 title(paste("Mean Fragment Size -", GRP, "\nBlue=Wt,  Red=Mt,  Pch=Locus (within Patient)"))
}
 

```




# STOP HERE

# References
<div id="refs"></div>

***
# Parameter settings
  * WRKDIR = `r WRKDIR`
  * FN = `r FN`
  * Scripts = Scripts
  * RUN DATE = `r date()`

```{r, echo=FALSE}
 sessionInfo()
```

```{r, echo=FALSE}
  knit_exit()
```




### ARCHIVAL CODE BELOW


<!-- To run
# nohup Rscript -e "knitr::knit2html('_M1A-exploreDELFI.Rmd')" > _M1A-exploreDELFI.log  &

# Or
# nohup Rscript -e "rmarkdown::render('_M1A-exploreDELFI.Rmd')" > _M1A-exploreDELFI.log  &

-->

